accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,165 @@
+
+Doing so would require source generators to carefully version and test across all framework versions.  This would increase testing cost for source generators and increase risk when shipping updates to source generators since they would apply to past, stable TargetFrameworks.  Conceptually source generators are lifting a substantial amount of runtime functionality out of a library and executing that at design time. This code is directly included in the user's application and contributes to runtime behavior.
+
+### Inplace servicing updates are good enough for the compiler, why not source generators?
+
+The compiler has a very stable 1:1 relationship between a user's source and the code it generates. As such the user has the ability to control the output of the compiler through their own usage.  The compiler's contribution of complexity to the user's assembly is very small and highly specified.
+ 
+Contrast this to source generators which can take a very small amount of user source to produce a large amount of generated code.  This large amount of generated code contains significant complexity which is loosely specified.  Its behavior is specific to the component it extends.  It's likely that we'd need to bugfix this code over time, and change it as we develop features in the components it extends.  It's also likely that we will learn new techniques and improvments to source generators over time, increasing the changes for bugfixes and features.
+
+Users expect stability for projects which do not change, for example a project targeting an LTS release. It is much more difficult to ensure this stability when we need to significantly components which contribute complexity to that project.
+
+### Is it bad to have multiple versions of a source generator possibly loaded by a repository?
+
+A repository may have projects targeting multiple frameworks.  These frameworks may contain different versions of the same source generator.  When building the repository this may require the same compiler server instance to load multiple copies of a source generator.  The same compiler would not likely load multiple copies of analyzers which do not ship side by side.  
+
+### Should all inbox source generators be tied to a shared framework?
+
+No, some source generators might be independent of libraries in a shared framework: these could go in a non-side-by-side location (eg: analyzers folder).  Other source generators may not contain functionality, but may instead just build some convenience API on top of existing public API.  These could be considered stable enough to place in a non-side-by-side location.
+
+### Should only source generators be shipped this way?  What about analyzers and code-fixes?
+
+Nothing is preventing analyzers and code fixes from shipping in the same manner.  If it's desirable for a such a component to ship with the framework rather than the compiler then that component may ship in this manner.
```

Comment:
I'd say we should strive to ship analyzers how we do today and only ship analyzers like source generators if the analyzer is specific to a generator:

1. For example, let's say one of our generators can be configured with custom attributes or naming convention. In that case we likely want an analyzers that can tell the developer when they messed up the configuration. Such an analyzer should likely ship in the same assembly as the generator because it's strongly tied.

2. A general purpose analyzer that is tied to a public API should likely ship like it does today because we generally don't need -- and don't want to -- tie those to a particular target framework.

Reply:
I was imagining an analyzer for a SQL connection string, or some other stringly typed API.  Any Analyzer trying to add better intellisense or design time validation of runtime functionality that's tied to features might make sense to ship this way.

Reply:
Clarified this a bit as

> Analyzers that need to validate source generator usage, or analyzers that are closely tied to runtime functionality may consider shipping side-by-side with the framework.